feat: Adding shaped recipes.

svaningelgem opened the pull request:
# Description

<!-- State the changes of the pull request below. -->

# Checklist

The following items should be checked before the pull request can be merged.

- [x] Code follows existing style.
- [x] Unit tests added (if applicable).

svaningelgem committed:
Adding shaped recipes.

svaningelgem committed:
removed comment

svaningelgem commented on src/test/java/org/mockbukkit/mockbukkit/inventory/RecipeManagerTest.java:
```diff
@@ -308,12 +308,6 @@ class ShapedRecipe
 					"CRIMSON_PLANKS, CRIMSON_DOOR",
 					"WARPED_PLANKS, WARPED_DOOR",
 					"COPPER_INGOT, COPPER_DOOR",
-					"EXPOSED_COPPER, EXPOSED_COPPER_DOOR",
```
These don't work in minecraft anyhow

svaningelgem committed:
Reducing complexity

svaningelgem committed:
Hopefully lower complexity

svaningelgem committed:
linting

svaningelgem committed:
Still reducing complexity

4everTheOne commented:
Hello! Thanks for your contribution! 
Looks like we have worked the same thing ☺️

https://github.com/MockBukkit/MockBukkit/pull/1396

I'll review it and see what we can retrieve from both solutions

svaningelgem commented:
Ooops. Mine seems simpler. But yours have more added tests & docs (and mine has full coverage 🎉). I'll stay off it to not duplicate our efforts any more :)

4everTheOne committed:
Add support for empty items and for mirrored recipes

4everTheOne committed:
Renamed variable

4everTheOne committed:
Improved comment

4everTheOne commented:
Hello again @svaningelgem! So, i agree with you, your solution is easier to understand than mine :)
I noticed that some use cases were missing, like when a shape has "null" ingredients like boats, and when the recipes is mirrored was not implemented.
I opened this PR at https://github.com/svaningelgem/MockBukkit/pull/1 to propose a solution for that. Let me know your thoughts :)

svaningelgem committed:
Merge pull request #1 from 4everTheOne/fix/add-flip-support

Add support for empty items and for mirrored recipes

svaningelgem committed:
Merge branch 'MockBukkit:v1.21' into feat-implementing-shaped-recipes

svaningelgem committed:
cleanup the code a little bit more

svaningelgem committed:
Improving performance (and coverage)

svaningelgem committed:
reorder modifiers

svaningelgem committed:
sonar issues

svaningelgem committed:
Silence warning during compilation

svaningelgem commented:
> Hello again @svaningelgem! So, i agree with you, your solution is easier to understand than mine :) I noticed that some use cases were missing, like when a shape has "null" ingredients like boats, and when the recipes is mirrored was not implemented. I opened this PR at [svaningelgem#1](https://github.com/svaningelgem/MockBukkit/pull/1) to propose a solution for that. Let me know your thoughts :)

I merged everything into this branch. I didn't really have any comments as it seemed straightforward.
I cleaned it up a little bit more though.
Also added full coverage (the mirror was sometimes not needed, and there always is at least 1 row).
